### ISSUE TITLE
pyright --verifytypes on pandas dev and numpy devs seems to hang

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#1424]
**Describe the bug**
Creating a dev environment using pandas master and numpy master and then doing pyright --verifytypes pandas gets no response after at least 30 minutes

**To Reproduce**
Create a pandas dev environment as described here:
https://pandas.pydata.org/docs/development/contributing.html#creating-a-development-environment
Create a py.typed file at the top level, and rename pandas/tests to pandas/_tests

Clone the numpy repo, build numpy and install it using pip into the same environment as described here:
https://numpy.org/devdocs/user/building.html#building-from-source

pyright --verifytypes pandas then doesn't respond.

**Expected behavior**
Results within a minute or so

**VS Code extension or command-line**
pyright command line version 1.1.105

**Additional context**
reference discussion here:
pandas-dev/pandas#28142 (comment)

and #1406

**Comment by erictraut** [comment]
I haven't been able to successfully build numpy, so I'm having trouble repro'ing this.

A few questions for you:
1. Did this problem start with the latest version of pyright? Or do you think it's likely that it occurred previously as well?
2. Does the problem occur for you if you run verifytypes on numpy alone? I'm trying to narrow down the problem. I suspect it's in numpy, not in pandas.
3. Are you deleting or renaming the test directory in pandas? As we previously discovered, this is a bunch of untyped code, and it takes pyright a long time to run verifytypes on it.

**Comment by Dr-Irv** [comment]
> 1. Did this problem start with the latest version of pyright? Or do you think it's likely that it occurred previously as well?

Was true with 1.1.104 and now 1.1.105

> 2. Does the problem occur for you if you run verifytypes on numpy alone? I'm trying to narrow down the problem. I suspect it's in numpy, not in pandas.

No, the fix in 1.1.105 now causes it to get through numpy, although it takes close to a minute (as compared to pandas, which takes about 5 seconds)

> 3. Are you deleting or renaming the test directory in pandas? As we previously discovered, this is a bunch of untyped code, and it takes pyright a long time to run verifytypes on it.

I'm renaming the `tests` directory to `_tests`

**Comment by Dr-Irv** [comment]
@erictraut numpy 1.20.0 was just released, so now you don't have to build numpy.  Just set up the pandas dev environment as described above, and do `conda install numpy=1.20.0 -c conda-forge` (or use `pip install numpy=1.20.0`), and when you do the type checking on pandas, it takes 68 seconds.

**Comment by erictraut** [comment]
It's harder than you might think to ignore anything due to importing one library from within another. A type that is completely undeclared and imported from another library gets an "Unknown" type, and pyright doesn't track where "Unknown" types come from. The good news is that pyright does track where classes, functions, and modules are declared, so it's relatively easy to filter out those types if they come from another package.

I've added a semi-hidden feature that will be in the next published release of pyright. To use the feature, append a bang ("!") to the end of the package name you are analyzing with the "--verifytypes" command.

Here's the output of `pyright --verifytypes pandas`:
```
Public symbols: 5089
  Symbols with unknown type: 4270
  Functions with missing docstring: 759
  Functions with missing default param: 3
  Classes with missing docstring: 106
Type completeness score: 16.1%

Completed in 51.551sec
```

And here's the output of `pyright --verifytypes pandas!`:
```
Public symbols: 5089
  Symbols with unknown type: 1232
    (Ignoring unknown types imported from other packages)
  Functions with missing docstring: 759
  Functions with missing default param: 3
  Classes with missing docstring: 106
Type completeness score: 75.8%

Completed in 4.936sec
```

As you can see, it runs _much_ faster and eliminates a lot of the "noise" you're seeing with partially-unknown types imported from other packages like numpy.

**Comment by erictraut** [comment]
This feature is included in Pyright 1.1.108, which I just published.

**Comment by Dr-Irv** [comment]
Thanks @erictraut .  This is great!
